### ISSUE TITLE
CUDA copy error in python backend on Jetson

### ISSUE THREAD

**Issue by KarelvdVrie** [triton-inference-server/server#2876]
**Description**
I am trying to parse the NMS output of a TensorRT engine using the Python backend on a Jetson TX2, but run into the following error: `try to use CUDA copy while GPU is not supported`.

When using the exact same models and ensemble network on an AMD64 platform it all runs without a problem.

The detection model was trained in the Transfer Learning Toolkit and converted using the tlt-converter. This all runs without any problems. However I cannot share this model, I hope this will not be a problem.

**Triton Information**
I am using the 21.04-py3 version on the AMD64 platform. 
And Docker image build with the TRITON v2.9.0 version from the release page using the nvcr.io/nvidia/l4t-base:r32.5.0, and with the Python backend build as in the documentation.

**To Reproduce**
Python model:
```

# Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#  * Neither the name of NVIDIA CORPORATION nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

import numpy as np
import sys
import json

# triton_python_backend_utils is available in every Triton Python model. You
# need to use this module to create inference requests and responses. It also
# contains some utility functions for extracting information from model_config
# and converting Triton input/output types to numpy types.
import triton_python_backend_utils as pb_utils

class TritonPythonModel:
    """Your Python model must use the same class name. Every Python model
    that is created must have "TritonPythonModel" as the class name.
    """

    def initialize(self, args):
        """`initialize` is called only once when the model is being loaded.
        Implementing `initialize` function is optional. This function allows
        the model to intialize any state associated with this model.
        Parameters
        ----------
        args : dict
          Both keys and values are strings. The dictionary keys and values are:
          * model_config: A JSON string containing the model configuration
          * model_instance_kind: A string containing model instance kind
          * model_instance_device_id: A string containing model instance device ID
          * model_repository: Model repository path
          * model_version: Model version
          * model_name: Model name
        """

        # You must parse model_config. JSON string is not parsed here
        self.model_config = model_config = json.loads(args['model_config'])

        # # Get OUTPUT configurations
        output_nms_config = pb_utils.get_output_config_by_name(
            model_config, "NMS")

        # # Convert Triton types to numpy types
        self.output_nms_dtype = pb_utils.triton_string_to_numpy(
            output_nms_config['data_type'])


    def execute(self, requests):
        """`execute` MUST be implemented in every Python model. `execute`
        function receives a list of pb_utils.InferenceRequest as the only
        argument. This function is called when an inference request is made
        for this model. Depending on the batching configuration (e.g. Dynamic
        Batching) used, `requests` may contain multiple requests. Every
        Python model, must create one pb_utils.InferenceResponse for every
        pb_utils.InferenceRequest in `requests`. If there is an error, you can
        set the error argument when creating a pb_utils.InferenceResponse
        Parameters
        ----------
        requests : list
          A list of pb_utils.InferenceRequest
        Returns
        -------
        list
          A list of pb_utils.InferenceResponse. The length of this list must
          be the same as `requests`
        """
        
        output_nms_dtype = self.output_nms_dtype
        responses = []

        # Every Python backend must iterate over everyone of the requests
        # and create a pb_utils.InferenceResponse for each of them.
        for request in requests:
            # Get INPUTS of this batch instance
            NMS = pb_utils.get_input_tensor_by_name(request, "NMS_INPUT").as_numpy()


            # Create output tensors. You need pb_utils.Tensor objects to create pb_utils.InferenceResponse.
            out_tensor_nms = pb_utils.Tensor("NMS", np.array(NMS).astype(output_nms_dtype))

            # Create InferenceResponse. You can set an error here in case
            # there was a problem with handling this inference request.
            # Below is an example of how you can set errors in inference
            # response:
            #
            # pb_utils.InferenceResponse(output_tensors=..., TritonError("An error occured"))
            inference_response = pb_utils.InferenceResponse(output_tensors=[out_tensor_nms])
            responses.append(inference_response)

        # You should return a list of pb_utils.InferenceResponse. Length
        # of this list must match the length of `requests` list.
        return responses

    def finalize(self):
        """`finalize` is called only once when the model is being unloaded.
        Implementing `finalize` function is OPTIONAL. This function allows
        the model to perform any necessary clean ups before exit.
        """
```

Python Config:
```
name: "PYTHON_NMS"
backend: "python"
default_model_filename: "PYTHON_NMS.py"

input [
  {
    name: "NMS_INPUT"
    data_type: TYPE_FP32
    dims: [ -1, 1, 200, 7 ]
  }
]


output [
  {
    name: "NMS"
    data_type: TYPE_FP32
    dims: [ -1, -1, -1, -1]
  }
]
instance_group [{ kind: KIND_CPU }]

```

Ensemble Network:
```
name: "Ensemble-_-OBJ_1-Detector--OBJ_1_NMS"
platform: "ensemble"
max_batch_size: 16

# Ensemble Inputs
input [
# General Image Input
	{
	  name: "IMG_IN"
	  data_type: TYPE_FP32
	  dims: [ 3, 640, 400 ]
	}
]

# Ensemble Outputs
output [
# OBJ 1 NMS Output
  {
    name: "NMS_PYTHON"
    data_type: TYPE_FP32
    dims: [ -1, -1, -1 ]
  }
]

# Ensemble steps
ensemble_scheduling {
  step [

    
  # OBJ 1
    {
      model_name: "Detector"
      model_version: -1
      
    # OBJ 1 Input
      input_map {
        key: "Input"
        value: "IMG_IN"
      }
      
    # OBJ 1 Output
      output_map {
        key: "NMS"
        value: "NMS"
      }
    },
  
  # Python NMS OBJ 1 
    {
      model_name: "PYTHON_NMS"
      model_version: -1
      
    # Python Inputs
      input_map {
        key: "NMS_INPUT"
        value: "NMS"
      }
      
    # Python Outputs
      output_map {
        key: "NMS"
        value: "NMS_PYTHON"
      }
    }
  ]
}


```

The TRITON server on the Jetson outputs the following log:
```
triton-inference-server | I0518 10:28:59.377489 1 http_server.cc:1229] HTTP request: 0 /v2/models/Ensemble-_-OBJ_1-Detector--OBJ_1_NMS
triton-inference-server | I0518 10:28:59.377885 1 model_repository_manager.cc:659] GetInferenceBackend() 'Ensemble-_-OBJ_1-Detector--OBJ_1_NMS' version -1
triton-inference-server | I0518 10:28:59.378006 1 model_repository_manager.cc:615] VersionStates() 'Ensemble-_-OBJ_1-Detector--OBJ_1_NMS'
triton-inference-server | I0518 10:28:59.379591 1 http_server.cc:1229] HTTP request: 0 /v2/models/Ensemble-_-OBJ_1-Detector--OBJ_1_NMS/config
triton-inference-server | I0518 10:28:59.379875 1 model_repository_manager.cc:659] GetInferenceBackend() 'Ensemble-_-OBJ_1-Detector--OBJ_1_NMS' version -1
triton-inference-server | I0518 10:29:01.250269 1 http_server.cc:1229] HTTP request: 2 /v2/models/Ensemble-_-OBJ_1-Detector--OBJ_1_NMS/infer
triton-inference-server | I0518 10:29:01.250418 1 model_repository_manager.cc:659] GetInferenceBackend() 'Ensemble-_-OBJ_1-Detector--OBJ_1_NMS' version -1
triton-inference-server | I0518 10:29:01.250467 1 model_repository_manager.cc:659] GetInferenceBackend() 'Ensemble-_-OBJ_1-Detector--OBJ_1_NMS' version -1
triton-inference-server | I0518 10:29:01.250664 1 infer_request.cc:497] prepared: [0x0x7ee00081e0] request id: , model: Ensemble-_-OBJ_1-Detector--OBJ_1_NMS, requested version: -1, actual version: 1, flags: 0x0, correlation id: 0, batch size: 6, priority: 0, timeout (us): 0
triton-inference-server | original inputs:
triton-inference-server | [0x0x7ee00034f8] input: IMG_IN, type: FP32, original shape: [6,3,640,400], batch + shape: [6,3,640,400], shape: [3,640,400]
triton-inference-server | override inputs:
triton-inference-server | inputs:
triton-inference-server | [0x0x7ee00034f8] input: IMG_IN, type: FP32, original shape: [6,3,640,400], batch + shape: [6,3,640,400], shape: [3,640,400]
triton-inference-server | original requested outputs:
triton-inference-server | NMS_PYTHON
triton-inference-server | requested outputs:
triton-inference-server | NMS_PYTHON
triton-inference-server | 
triton-inference-server | I0518 10:29:01.250796 1 model_repository_manager.cc:659] GetInferenceBackend() 'Detector' version -1
triton-inference-server | I0518 10:29:01.250843 1 model_repository_manager.cc:659] GetInferenceBackend() 'PYTHON_NMS' version -1
triton-inference-server | I0518 10:29:01.250927 1 infer_request.cc:497] prepared: [0x0x7ee000f2e0] request id: , model: Detector, requested version: -1, actual version: 4, flags: 0x0, correlation id: 0, batch size: 6, priority: 0, timeout (us): 0
triton-inference-server | original inputs:
triton-inference-server | [0x0x7ee0001f78] input: Input, type: FP32, original shape: [6,3,640,400], batch + shape: [6,3,640,400], shape: [3,640,400]
triton-inference-server | override inputs:
triton-inference-server | inputs:
triton-inference-server | [0x0x7ee0001f78] input: Input, type: FP32, original shape: [6,3,640,400], batch + shape: [6,3,640,400], shape: [3,640,400]
triton-inference-server | original requested outputs:
triton-inference-server | NMS
triton-inference-server | requested outputs:
triton-inference-server | NMS
triton-inference-server | 
triton-inference-server | I0518 10:29:01.251267 1 plan_backend.cc:2447] Running Detector_0_gpu0 with 1 requests
triton-inference-server | I0518 10:29:01.251436 1 plan_backend.cc:3378] Optimization profile default [0] is selected for Detector_0_gpu0
triton-inference-server | I0518 10:29:01.274285 1 plan_backend.cc:2869] Context with profile default [0] is being executed for Detector_0_gpu0
triton-inference-server | I0518 10:29:01.287877 1 infer_response.cc:139] add response output: output: NMS, type: FP32, shape: [6,1,200,7]
triton-inference-server | I0518 10:29:01.287997 1 ensemble_scheduler.cc:509] Internal response allocation: NMS, size 33600, addr 0xf00e80000, memory type 2, type id 0
triton-inference-server | I0518 10:29:01.688509 1 ensemble_scheduler.cc:524] Internal response release: size 33600, addr 0xf00e80000
triton-inference-server | I0518 10:29:01.688760 1 infer_request.cc:497] prepared: [0x0x7eac001040] request id: , model: PYTHON_NMS, requested version: -1, actual version: 1, flags: 0x0, correlation id: 0, batch size: 0, priority: 0, timeout (us): 0
triton-inference-server | original inputs:
triton-inference-server | [0x0x7eac0012c8] input: NMS_INPUT, type: FP32, original shape: [6,1,200,7], batch + shape: [6,1,200,7], shape: [6,1,200,7]
triton-inference-server | override inputs:
triton-inference-server | inputs:
triton-inference-server | [0x0x7eac0012c8] input: NMS_INPUT, type: FP32, original shape: [6,1,200,7], batch + shape: [6,1,200,7], shape: [6,1,200,7]
triton-inference-server | original requested outputs:
triton-inference-server | NMS
triton-inference-server | requested outputs:
triton-inference-server | NMS
triton-inference-server | 
triton-inference-server | I0518 10:29:01.689303 1 python.cc:351] model PYTHON_NMS, instance PYTHON_NMS_0, executing 1 requests
triton-inference-server | E0518 10:29:01.691989 1 python.cc:645] Request 0: failed to create output response
triton-inference-server | I0518 10:29:01.692123 1 python.cc:688] TRITONBACKEND_ModelInstanceExecute: model instance name PYTHON_NMS_0 released 1 requests
```

**Expected behavior**
I expected the python backend to work the same as on an AMD64 platform. Thus that the NMS output of the detection network would be accessible by the Python backend within the CPU space without any additional code.

**Comment by Tabrizian** [comment]
@KarelvdVrie Thanks for filing the bug, Where is the error that you have shared in the first line of the output is printed? Is it printed in the client logs or the server logs? I can't find that line in the logs that you have posted. Also, we recently fixed a bug with regard to the output buffers in Python backend. Can you try 21.05 to see whether it fixes your issue?

**Comment by KarelvdVrie** [comment]
The output is printed in the client logs. Below the full traceback:
```
Traceback (most recent call last):
  File "/Development/Client.py", line 870, in <module>
    main(FLAGS, INFER, MQTT_CLIENT)
  File "/Development/Client.py", line 296, in main
    batch_result = INFER["triton_client"].infer(FLAGS.model_name, inputs, outputs=outputs)
  File "/usr/local/lib/python3.9/site-packages/tritonclient/http/__init__.py", line 1112, in infer
    _raise_if_error(response)
  File "/usr/local/lib/python3.9/site-packages/tritonclient/http/__init__.py", line 63, in _raise_if_error
    raise error
tritonclient.utils.InferenceServerException: in ensemble 'Ensemble-_-OBJ_1-Detector--OBJ_1_NMS', NMS_INPUT: try to use CUDA copy while GPU is not supported
```
Due to the NSM_INPUT it seems to me that error originates in the Python code on the server side.

I am not using shared memory to communicate between the client and server. 

I will try 21.05 hopefully later this week. I will need to build this from source, however the build for Jetpack is still "Under Construction". So it might be easier if I wait for the next release to come out.

**Comment by CoderHam** [comment]
@KarelvdVrie can you share the build flags /steps you used to build your python backend? The error seems to indicate that you built with TRITON_ENABLE_GPU OFF

**Comment by KarelvdVrie** [comment]
I use the following Dockerfile to build the image:
```
FROM nvcr.io/nvidia/l4t-base:r32.5.0

ENV CMAKE_VERSION=3.20.2

ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get update && \
    apt-get install -y --no-install-recommends \
        software-properties-common \
        autoconf \
        automake \
        build-essential \
        git \
        libb64-dev \
        libre2-dev \
        libssl-dev \
        libtool \
        libboost-dev \
        libcurl4-openssl-dev \
        rapidjson-dev \
        patchelf \
        zlib1g-dev

WORKDIR /Workdir/
# Installing recent cmake version

RUN wget https://github.com/Kitware/CMake/releases/download/v$CMAKE_VERSION/cmake-$CMAKE_VERSION-linux-aarch64.tar.gz \
	&& tar xvf cmake-$CMAKE_VERSION-linux-aarch64.tar.gz 

RUN ln -s /Workdir/cmake-$CMAKE_VERSION-linux-aarch64/bin/cmake /usr/bin/cmake

ADD tritonserver2.9.0-jetpack4.5.tgz /Workdir/TRITON/


# Build Python backend
ENV BACKEND_TAG=r21.04
ENV TRITON_PATH=/Workdir/TRITON/

WORKDIR /Workdir/Python_Backend

RUN apt-get update && \
    apt-get install -y --no-install-recommends \
        python3-dev \
        python3-pip \
        rapidjson-dev \
        git
        
RUN pip3 install -U pip setuptools wheel
RUN pip3 install -U Cython

# Needed for python3.6 keeping back numpy version 1.19. import will fail
ENV OPENBLAS_CORETYPE=ARMV8 
RUN pip3 install -U numpy

RUN git clone --branch $BACKEND_TAG https://github.com/triton-inference-server/python_backend .

# Fix for the filesystem error, something with the c language version 11 vs 17
RUN sed -i 's/include <filesystem>/include <experimental\/filesystem>/g' src/python.cc

RUN mkdir build \
	&& cd build \
	&& cmake -DCMAKE_INSTALL_PREFIX:PATH=$TRITON_PATH -DTRITON_BACKEND_REPO_TAG=$BACKEND_TAG -DTRITON_COMMON_REPO_TAG=$BACKEND_TAG .. 
	
RUN cd build \
	&& ls ../ \
	&& cat ../src/python.cc \
	&& cat ../CMakeLists.txt \
	&& make install
	
WORKDIR /Workdir/
```

I indeed do not use the TRITON_ENABLE_GPU. It is not mentioned on the [Python Backend](https://github.com/triton-inference-server/python_backend) repository. I will try to enable that flag and come back to you.

**Comment by KarelvdVrie** [comment]
The Python backend cannot be build using the TRITON_ENABLE_GPU flag  `CMake Error: Parse error in command line argument: TRITON_ENABLE_GPU`  and I take it the Jetson releases are build using the flag. So how would I go about this?

**Comment by CoderHam** [comment]
This is odd. TRITON_ENABLE_GPU is indeed added as a cmake flag option. Did you use it as `-DTRITON_ENABLE_GPU=ON`
https://github.com/triton-inference-server/python_backend/blob/r21.05/CMakeLists.txt#L40

**Comment by KarelvdVrie** [comment]
Ah I had a small typo in it, now it is building, will let you know if this fixes the problem.

**Comment by KarelvdVrie** [comment]
The build was succesfull however running the resulting image is not. For some reason the backend api version does not support the backend.

```
I0519 08:11:49.106210 1 shared_library.cc:44] OpenLibraryHandle: /Workdir/TRITON/backends/tensorflow1/libtriton_tensorflow1.so
2021-05-19 08:11:49.673614: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.10.2
I0519 08:11:49.796025 1 tensorflow.cc:2167] TRITONBACKEND_Initialize: tensorflow
I0519 08:11:49.796131 1 tensorflow.cc:2180] Triton TRITONBACKEND API version: 1.0
I0519 08:11:49.796157 1 tensorflow.cc:2186] 'tensorflow' TRITONBACKEND API version: 1.0
I0519 08:11:49.796176 1 tensorflow.cc:2207] backend configuration:
{}
I0519 08:11:49.939692 1 pinned_memory_manager.cc:206] Pinned memory pool is created at '0x101070000' with size 268435456
I0519 08:11:49.940058 1 cuda_memory_manager.cc:103] CUDA memory pool is created on device 0 with size 67108864
I0519 08:11:49.942208 1 backend_factory.h:44] Create TritonBackendFactory
I0519 08:11:49.942294 1 plan_backend_factory.cc:48] Create PlanBackendFactory
I0519 08:11:49.942314 1 plan_backend_factory.cc:55] Registering TensorRT Plugins
I0519 08:11:49.942427 1 logging.cc:52] Registered plugin creator - ::BatchTilePlugin_TRT version 1
I0519 08:11:49.942504 1 logging.cc:52] Registered plugin creator - ::BatchedNMS_TRT version 1
I0519 08:11:49.942538 1 logging.cc:52] Registered plugin creator - ::BatchedNMSDynamic_TRT version 1
I0519 08:11:49.942573 1 logging.cc:52] Registered plugin creator - ::CoordConvAC version 1
I0519 08:11:49.942600 1 logging.cc:52] Registered plugin creator - ::CropAndResize version 1
I0519 08:11:49.942630 1 logging.cc:52] Registered plugin creator - ::DetectionLayer_TRT version 1
I0519 08:11:49.942660 1 logging.cc:52] Registered plugin creator - ::FlattenConcat_TRT version 1
I0519 08:11:49.942687 1 logging.cc:52] Registered plugin creator - ::GenerateDetection_TRT version 1
I0519 08:11:49.942722 1 logging.cc:52] Registered plugin creator - ::GridAnchor_TRT version 1
I0519 08:11:49.942749 1 logging.cc:52] Registered plugin creator - ::GridAnchorRect_TRT version 1
I0519 08:11:49.942783 1 logging.cc:52] Registered plugin creator - ::InstanceNormalization_TRT version 1
I0519 08:11:49.942810 1 logging.cc:52] Registered plugin creator - ::MultilevelCropAndResize_TRT version 1
I0519 08:11:49.942845 1 logging.cc:52] Registered plugin creator - ::MultilevelProposeROI_TRT version 1
I0519 08:11:49.942872 1 logging.cc:52] Registered plugin creator - ::NMS_TRT version 1
I0519 08:11:49.942903 1 logging.cc:52] Registered plugin creator - ::Normalize_TRT version 1
I0519 08:11:49.942937 1 logging.cc:52] Registered plugin creator - ::PriorBox_TRT version 1
I0519 08:11:49.943033 1 logging.cc:52] Registered plugin creator - ::ProposalLayer_TRT version 1
I0519 08:11:49.943072 1 logging.cc:52] Registered plugin creator - ::Proposal version 1
I0519 08:11:49.943106 1 logging.cc:52] Registered plugin creator - ::PyramidROIAlign_TRT version 1
I0519 08:11:49.943141 1 logging.cc:52] Registered plugin creator - ::Region_TRT version 1
I0519 08:11:49.943173 1 logging.cc:52] Registered plugin creator - ::Reorg_TRT version 1
I0519 08:11:49.943203 1 logging.cc:52] Registered plugin creator - ::ResizeNearest_TRT version 1
I0519 08:11:49.943237 1 logging.cc:52] Registered plugin creator - ::RPROI_TRT version 1
I0519 08:11:49.943266 1 logging.cc:52] Registered plugin creator - ::SpecialSlice_TRT version 1
I0519 08:11:49.943314 1 custom_backend_factory.cc:46] Create CustomBackendFactory
I0519 08:11:49.943340 1 ensemble_backend_factory.cc:47] Create EnsembleBackendFactory
I0519 08:11:49.946347 1 autofill.cc:138] TensorFlow SavedModel autofill: Internal: unable to autofill for 'PYTHON_NMS', unable to find savedmodel directory named 'model.savedmodel'
I0519 08:11:49.946533 1 autofill.cc:151] TensorFlow GraphDef autofill: Internal: unable to autofill for 'PYTHON_NMS', unable to find graphdef file named 'model.graphdef'
E0519 08:11:54.697848 1 logging.cc:43] coreReadArchive.cpp (31) - Serialization Error in verifyHeader: 0 (Magic tag does not match)
E0519 08:11:54.698254 1 logging.cc:43] INVALID_STATE: std::exception
E0519 08:11:54.698329 1 logging.cc:43] INVALID_CONFIG: Deserialize the cuda engine failed.
I0519 08:11:54.698392 1 autofill.cc:209] TensorRT autofill: Internal: unable to autofill for 'PYTHON_NMS', unable to find a compatible plan file.
W0519 08:11:54.698427 1 autofill.cc:248] Proceeding with simple config for now
I0519 08:11:54.698452 1 model_config_utils.cc:637] autofilled config: name: "PYTHON_NMS"
input {
  name: "NMS_INPUT"
  data_type: TYPE_FP32
  dims: -1
  dims: 1
  dims: 200
  dims: 7
}
output {
  name: "NMS"
  data_type: TYPE_FP32
  dims: -1
  dims: -1
  dims: -1
  dims: -1
}
instance_group {
  kind: KIND_CPU
}
default_model_filename: "PYTHON_NMS.py"
backend: "python"

I0519 08:11:54.698823 1 model_repository_manager.cc:770] AsyncLoad() 'PYTHON_NMS'
I0519 08:11:54.699181 1 model_repository_manager.cc:1009] TriggerNextAction() 'PYTHON_NMS' version 1: 1
I0519 08:11:54.699227 1 model_repository_manager.cc:1047] Load() 'PYTHON_NMS' version 1
I0519 08:11:54.699246 1 model_repository_manager.cc:1066] loading: PYTHON_NMS:1
I0519 08:11:54.800577 1 model_repository_manager.cc:1126] CreateInferenceBackend() 'PYTHON_NMS' version 1
I0519 08:11:54.801023 1 shared_library.cc:44] OpenLibraryHandle: /Workdir/TRITON/backends/python/libtriton_python.so
I0519 08:11:54.802748 1 python.cc:998] 'python' TRITONBACKEND API version: 1.2
I0519 08:11:54.802839 1 triton_backend_manager.cc:92] unloading backend 'python'
I0519 08:11:54.802872 1 python.cc:1104] TRITONBACKEND_Finalize: Start
I0519 08:11:54.802900 1 python.cc:1109] TRITONBACKEND_Finalize: End
E0519 08:11:54.803470 1 model_repository_manager.cc:1243] failed to load 'PYTHON_NMS' version 1: Unsupported: Triton backend API version does not support this backend
I0519 08:11:54.803540 1 model_repository_manager.cc:1009] TriggerNextAction() 'PYTHON_NMS' version 1: 0
I0519 08:11:54.803579 1 model_repository_manager.cc:1024] no next action, trigger OnComplete()
I0519 08:11:54.803828 1 model_repository_manager.cc:615] VersionStates() 'PYTHON_NMS'
I0519 08:11:54.804678 1 model_repository_manager.cc:615] VersionStates() 'PYTHON_NMS'
I0519 08:11:54.804873 1 server.cc:504] 
+------------------+------+
| Repository Agent | Path |
+------------------+------+
+------------------+------+

I0519 08:11:54.805066 1 server.cc:543] 
+------------+---------------------------------------------------------------+--------+
| Backend    | Path                                                          | Config |
+------------+---------------------------------------------------------------+--------+
| tensorrt   | <built-in>                                                    | {}     |
| tensorflow | /Workdir/TRITON/backends/tensorflow1/libtriton_tensorflow1.so | {}     |
+------------+---------------------------------------------------------------+--------+

I0519 08:11:54.805101 1 model_repository_manager.cc:591] BackendStates()
I0519 08:11:54.805250 1 server.cc:586] 
+------------+---------+------------------------------------------------------------------------------------+
| Model      | Version | Status                                                                             |
+------------+---------+------------------------------------------------------------------------------------+
| PYTHON_NMS | 1       | UNAVAILABLE: Unsupported: Triton backend API version does not support this backend |
+------------+---------+------------------------------------------------------------------------------------+

I0519 08:11:54.805665 1 tritonserver.cc:1658] 
+----------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Option                           | Value                                                                                                                                                                                  |
+----------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| server_id                        | triton                                                                                                                                                                                 |
| server_version                   | 2.9.0                                                                                                                                                                                  |
| server_extensions                | classification sequence model_repository model_repository(unload_dependents) schedule_policy model_configuration system_shared_memory cuda_shared_memory binary_tensor_data statistics |
| model_repository_path[0]         | /Workdir/Models                                                                                                                                                                        |
| model_control_mode               | MODE_NONE                                                                                                                                                                              |
| strict_model_config              | 0                                                                                                                                                                                      |
| pinned_memory_pool_byte_size     | 268435456                                                                                                                                                                              |
| cuda_memory_pool_byte_size{0}    | 67108864                                                                                                                                                                               |
| min_supported_compute_capability | 5.3                                                                                                                                                                                    |
| strict_readiness                 | 1                                                                                                                                                                                      |
| exit_timeout                     | 30                                                                                                                                                                                     |
+----------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

E0519 08:11:54.805731 1 tritonserver.cc:1667] Internal: failed to load all models
I0519 08:11:54.808182 1 grpc_server.cc:225] Ready for RPC 'ServerLive', 0
I0519 08:11:54.808268 1 grpc_server.cc:225] Ready for RPC 'ServerReady', 0
I0519 08:11:54.808303 1 grpc_server.cc:225] Ready for RPC 'ModelReady', 0
I0519 08:11:54.808332 1 grpc_server.cc:225] Ready for RPC 'ServerMetadata', 0
I0519 08:11:54.808397 1 grpc_server.cc:225] Ready for RPC 'ModelMetadata', 0
I0519 08:11:54.808432 1 grpc_server.cc:225] Ready for RPC 'ModelConfig', 0
I0519 08:11:54.808458 1 grpc_server.cc:225] Ready for RPC 'ModelStatistics', 0
I0519 08:11:54.808513 1 grpc_server.cc:225] Ready for RPC 'SystemSharedMemoryStatus', 0
I0519 08:11:54.808548 1 grpc_server.cc:225] Ready for RPC 'SystemSharedMemoryRegister', 0
I0519 08:11:54.808592 1 grpc_server.cc:225] Ready for RPC 'SystemSharedMemoryUnregister', 0
I0519 08:11:54.808620 1 grpc_server.cc:225] Ready for RPC 'CudaSharedMemoryStatus', 0
I0519 08:11:54.808692 1 grpc_server.cc:225] Ready for RPC 'CudaSharedMemoryRegister', 0
I0519 08:11:54.808986 1 grpc_server.cc:225] Ready for RPC 'CudaSharedMemoryUnregister', 0
I0519 08:11:54.809029 1 grpc_server.cc:225] Ready for RPC 'RepositoryIndex', 0
I0519 08:11:54.809059 1 grpc_server.cc:225] Ready for RPC 'RepositoryModelLoad', 0
I0519 08:11:54.809129 1 grpc_server.cc:225] Ready for RPC 'RepositoryModelUnload', 0
I0519 08:11:54.809213 1 grpc_server.cc:416] Thread started for CommonHandler
I0519 08:11:54.809667 1 grpc_server.cc:3117] New request handler for ModelInferHandler, 1
I0519 08:11:54.809756 1 grpc_server.cc:2176] Thread started for ModelInferHandler
I0519 08:11:54.810065 1 grpc_server.cc:3467] New request handler for ModelStreamInferHandler, 3
I0519 08:11:54.810156 1 grpc_server.cc:2176] Thread started for ModelStreamInferHandler
I0519 08:11:54.810189 1 grpc_server.cc:4028] Started GRPCInferenceService at 0.0.0.0:8001
I0519 08:11:54.811135 1 http_server.cc:2761] Started HTTPService at 0.0.0.0:8000
I0519 08:11:54.854803 1 http_server.cc:2780] Started Metrics Service at 0.0.0.0:8002

```

What surprises me is that in one of the first lines the following is logged `I0519 08:11:49.796131 1 tensorflow.cc:2180] Triton TRITONBACKEND API version: 1.0`. 
But that later when trying to load the Python backend it logs `I0519 08:11:54.802748 1 python.cc:998] 'python' TRITONBACKEND API version: 1.2`.
Did I miss something when setting the BACKEND_TAG?

I am using the Dockerfile from above with the addition of the gpu flag.

**Comment by tanmayv25** [comment]
We have updated the backend API on our `main` branch. Please use the following cmake option while building python backend: 
`-DTRITON_CORE_REPO_TAG=r21.05`
https://github.com/triton-inference-server/python_backend/blob/r21.05/CMakeLists.txt#L45

**Comment by KarelvdVrie** [comment]
I take it I will need to build the r21.05 version of the TRITON server as well. Which comes back to my question where I can find documentation on how to do this for the Jetson platforms?
Using the r21.05 version of the Python backend with the prebiuld r21.04 version of TRITON does not work.

**Comment by KarelvdVrie** [comment]
Ah using the following it builds and loads the Python model: `cmake -DCMAKE_INSTALL_PREFIX:PATH=/Workdir/TRITON/ -DTRITON_BACKEND_REPO_TAG=r21.04 -DTRITON_ENABLE_GPU=ON -DTRITON_COMMON_REPO_TAG=r21.04 -DTRITON_CORE_REPO_TAG=r21.05 .. `

This also fixes my problem, thank you for the responses!

**Comment by bpfliegel** [comment]
> Ah using the following it builds and loads the Python model: `cmake -DCMAKE_INSTALL_PREFIX:PATH=/Workdir/TRITON/ -DTRITON_BACKEND_REPO_TAG=r21.04 -DTRITON_ENABLE_GPU=ON -DTRITON_COMMON_REPO_TAG=r21.04 -DTRITON_CORE_REPO_TAG=r21.05 .. `
> 
> This also fixes my problem, thank you for the responses!

Could you be so kind to share your final docker file? I was struggling with the same problem. I would be nice to have something rock solid as a start :) Thanks very much! Balint

**Comment by KarelvdVrie** [comment]
You basically use the Dockerfile described earlier and update the cmake command:

```
FROM nvcr.io/nvidia/l4t-base:r32.5.0

ENV CMAKE_VERSION=3.20.2

ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get update && \
    apt-get install -y --no-install-recommends \
        software-properties-common \
        autoconf \
        automake \
        build-essential \
        git \
        libb64-dev \
        libre2-dev \
        libssl-dev \
        libtool \
        libboost-dev \
        libcurl4-openssl-dev \
        rapidjson-dev \
        patchelf \
        zlib1g-dev

WORKDIR /Workdir/
# Installing recent cmake version

RUN wget https://github.com/Kitware/CMake/releases/download/v$CMAKE_VERSION/cmake-$CMAKE_VERSION-linux-aarch64.tar.gz \
	&& tar xvf cmake-$CMAKE_VERSION-linux-aarch64.tar.gz 

RUN ln -s /Workdir/cmake-$CMAKE_VERSION-linux-aarch64/bin/cmake /usr/bin/cmake

ADD tritonserver2.9.0-jetpack4.5.tgz /Workdir/TRITON/


# Build Python backend
ENV BACKEND_TAG=r21.04
ENV TRITON_PATH=/Workdir/TRITON/

WORKDIR /Workdir/Python_Backend

RUN apt-get update && \
    apt-get install -y --no-install-recommends \
        python3-dev \
        python3-pip \
        rapidjson-dev \
        git
        
RUN pip3 install -U pip setuptools wheel
RUN pip3 install -U Cython

# Needed for python3.6 keeping back numpy version 1.19. import will fail
ENV OPENBLAS_CORETYPE=ARMV8 
RUN pip3 install -U numpy

RUN git clone --branch $BACKEND_TAG https://github.com/triton-inference-server/python_backend .

# Fix for the filesystem error, something with the c language version 11 vs 17
RUN sed -i 's/include <filesystem>/include <experimental\/filesystem>/g' src/python.cc

RUN mkdir build \
	&& cd build \
	&& cmake -DCMAKE_INSTALL_PREFIX:PATH=$TRITON_PATH -DTRITON_BACKEND_REPO_TAG=$BACKEND_TAG -DTRITON_COMMON_REPO_TAG=$BACKEND_TAG -DTRITON_ENABLE_GPU=ON -DTRITON_CORE_REPO_TAG=r21.05 ..
	
RUN cd build \
	&& ls ../ \
	&& cat ../src/python.cc \
	&& cat ../CMakeLists.txt \
	&& make install
	
WORKDIR /Workdir/
```

**Comment by bpfliegel** [comment]
Thanks mate, got it working.

I still have to apply this first to be able to build, not just run (newbie with dockers on Jetsons): https://stackoverflow.com/a/61737404

And this worked for me at the end (commands at the end for testing):

`FROM nvcr.io/nvidia/l4t-base:r32.5.0

ENV CMAKE_VERSION=3.20.2

ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get update && \
    apt-get install -y --no-install-recommends \
        software-properties-common \
        autoconf \
        automake \
        build-essential \
        git \
        libb64-dev \
        libre2-dev \
        libssl-dev \
        libtool \
        libboost-dev \
        libcurl4-openssl-dev \
        rapidjson-dev \
        patchelf \
        zlib1g-dev

WORKDIR /Workdir/
# Installing recent cmake version

RUN wget https://github.com/Kitware/CMake/releases/download/v$CMAKE_VERSION/cmake-$CMAKE_VERSION-linux-aarch64.tar.gz \
	&& tar xvf cmake-$CMAKE_VERSION-linux-aarch64.tar.gz 

RUN ln -s /Workdir/cmake-$CMAKE_VERSION-linux-aarch64/bin/cmake /usr/bin/cmake

#ADD tritonserver2.9.0-jetpack4.5.tgz /Workdir/TRITON/
RUN wget https://github.com/triton-inference-server/server/releases/download/v2.9.0/tritonserver2.9.0-jetpack4.5.tgz
RUN mkdir -p /Workdir/TRITON/
RUN tar -xzvf tritonserver2.9.0-jetpack4.5.tgz -C /Workdir/TRITON/


# Build Python backend
ENV BACKEND_TAG=r21.04
ENV TRITON_PATH=/Workdir/TRITON/

WORKDIR /Workdir/Python_Backend

RUN apt-get update && \
    apt-get install -y --no-install-recommends \
        python3-dev \
        python3-pip \
        rapidjson-dev \
        git
        
RUN pip3 install -U pip setuptools wheel
RUN pip3 install -U Cython

# Needed for python3.6 keeping back numpy version 1.19. import will fail
ENV OPENBLAS_CORETYPE=ARMV8 
RUN pip3 install -U numpy

RUN git clone --branch $BACKEND_TAG https://github.com/triton-inference-server/python_backend .

# Fix for the filesystem error, something with the c language version 11 vs 17
RUN sed -i 's/include <filesystem>/include <experimental\/filesystem>/g' src/python.cc

RUN mkdir build \
	&& cd build \
	&& cmake -DCMAKE_INSTALL_PREFIX:PATH=$TRITON_PATH -DTRITON_BACKEND_REPO_TAG=$BACKEND_TAG -DTRITON_COMMON_REPO_TAG=$BACKEND_TAG -DTRITON_ENABLE_GPU=ON -DTRITON_CORE_REPO_TAG=r21.05 ..
	
RUN cd build \
	&& ls ../ \
	&& cat ../src/python.cc \
	&& cat ../CMakeLists.txt \
	&& make install
	
WORKDIR /Workdir/

RUN cd /tmp && git clone https://github.com/triton-inference-server/python_backend -b $BACKEND_TAG 
RUN mkdir -p /models/add_sub/1/
RUN cp /tmp/python_backend/examples/add_sub/model.py /models/add_sub/1/model.py
RUN cp /tmp/python_backend/examples/add_sub/config.pbtxt /models/add_sub/config.pbtxt

# Commands
# Build it as: triton:1.0
# Bash into it: docker run --runtime nvidia -it triton:1.0 /bin/bash
# Test: /Workdir/TRITON/bin/tritonserver --model-repository=/models --backend-directory=/Workdir/TRITON/backends --backend-config=python,python-runtime=/usr/bin/python

`

**Comment by bpfliegel** [comment]
One note: if you install gcc-8  and g++-8 and apply:
```
RUN update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-7 700 --slave /usr/bin/g++ g++ /usr/bin/g++-7
RUN update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-8 800 --slave /usr/bin/g++ g++ /usr/bin/g++-8
```
you can build normally, without the
```
RUN sed -i 's/include <filesystem>/include <experimental\/filesystem>/g' src/python.cc
```
hack.

Thanks again!
